src/librustc/ty/context.rs

```diff
@@ -1089,6 +1094,52 @@ impl<'tcx> TyCtxt<'tcx> {
         &self.hir_map
     }
 
+    /// In interpreter mode, gets the `DefId` of the user fn (a closure marked by an attribute named
+    /// `rustc_interp_user_fn`). This first time this method is called, that fn is located in the
+    /// crate HIR, and is cached for subsequent calls.
+    #[inline]
+    pub fn get_interp_user_fn(&self) -> DefId {
```

Comment:
Why is this not a query? It sounds like that would be exactly what we want here -- re:caching and such?

Reply:
Yes, it probably should be! I have no experience creating queries though. But if someone could either instruct me how to, or make a commit to this branch querifying this, then great.

Reply:
Yeah, sure!

You'll want to edit `src/librustc/query/mod.rs` -- probably towards the very end in the Other category you can add a query like `query interp_user_fn(_: CrateNum) -> DefId` or so, maybe with an `eval_always` flag -- look around in that file -- and after that's done, you make a free standing function somewhere, and for example in ty/context.rs (at the end) you can do `providers.interp_user_fn = interp_user_fn` or so (or a closure, like you can see there).

I think that'll be pretty much it -- you can just assert that the crate num you're passed is the local crate; in theory you could also add a cross-crate impl but that I believe is not necessary for this, right?

Reply:
Okay thanks!